File: core/src/main/java/com/canyon/commons/UUIDUtils.java
package com.canyon.commons;

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.UUID;

public class UUIDUtils {

	public final static UUID UUID_EMPTY = new UUID(0, 0);

	public static UUID fromStringWithHyphens(String str) {
		return UUID.fromString(str);
	}

	public static UUID fromStringWhitoutHyphens(String str) {
		if (str.length() != 32) {
			throw new IllegalArgumentException("Invalid UUID string: " + str);
		}
		String s1 = "0x" + str.substring(0, 8);
		String s2 = "0x" + str.substring(9, 12);
		String s3 = "0x" + str.substring(13, 16);
		String s4 = "0x" + str.substring(17, 20);
		String s5 = "0x" + str.substring(21, 32);

		long mostSigBits = Long.decode(s1).longValue();
		mostSigBits <<= 16;
		mostSigBits |= Long.decode(s2).longValue();
		mostSigBits <<= 16;
		mostSigBits |= Long.decode(s3).longValue();

		long leastSigBits = Long.decode(s4).longValue();
		leastSigBits <<= 48;
		leastSigBits |= Long.decode(s5).longValue();

		return new UUID(mostSigBits, leastSigBits);
	}

	public static UUID fromString(String str) {
		if (str.indexOf((char) '-') >= 0) {
			return fromStringWithHyphens(str);
		} else {
			return fromStringWhitoutHyphens(str);
		}
	}

	public static String toString(UUID uuid) {
		return uuid.toString();
	}

	public static String toStringWhitoutHyphens(UUID uuid) {
		long msb = uuid.getMostSignificantBits();
		long lsb = uuid.getLeastSignificantBits();
		return (digits(msb >> 32, 8) + digits(msb >> 16, 4) + digits(msb, 4)
				+ digits(lsb >> 48, 4) + digits(lsb, 12));
	}

	private static String digits(long val, int digits) {
		long hi = 1L << (digits * 4);
		return Long.toHexString(hi | (val & (hi - 1))).substring(1);
	}

	public static UUID fromByte(byte[] data) {
		if (data.length != 16) {
			throw new IllegalArgumentException("Invalid UUID byte[]");
		}

		long msb = 0;
		long lsb = 0;
		for (int i = 0; i < 8; i++)
			msb = (msb << 8) | (data[i] & 0xff);
		for (int i = 8; i < 16; i++)
			lsb = (lsb << 8) | (data[i] & 0xff);

		return new UUID(msb, lsb);
	}

	public static byte[] toByte(UUID uuid) {
		ByteArrayOutputStream ba = new ByteArrayOutputStream(16);
		DataOutputStream da = new DataOutputStream(ba);
		try {
			da.writeLong(uuid.getMostSignificantBits());
			da.writeLong(uuid.getLeastSignificantBits());
		} catch (IOException e) {
			e.printStackTrace();
		}
		return ba.toByteArray();
	}
}

File: commons/src/main/kotlin/com/canyon/commons/UUIDUtils.kt
package com.canyon.commons

import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.util.*
import kotlin.experimental.and

object UUIDUtils {

    val UUID_EMPTY = UUID(0, 0)

    fun fromStringWithHyphens(str: String): UUID {
        return UUID.fromString(str)
    }

    fun fromStringWhitoutHyphens(str: String): UUID {
        if (str.length != 32) {
            throw IllegalArgumentException("Invalid UUID string: $str")
        }
        val s1 = "0x" + str.substring(0, 8)
        val s2 = "0x" + str.substring(9, 12)
        val s3 = "0x" + str.substring(13, 16)
        val s4 = "0x" + str.substring(17, 20)
        val s5 = "0x" + str.substring(21, 32)

        var mostSigBits = java.lang.Long.decode(s1).toLong()
        mostSigBits = mostSigBits shl 16
        mostSigBits = mostSigBits or java.lang.Long.decode(s2).toLong()
        mostSigBits = mostSigBits shl 16
        mostSigBits = mostSigBits or java.lang.Long.decode(s3).toLong()

        var leastSigBits = java.lang.Long.decode(s4).toLong()
        leastSigBits = leastSigBits shl 48
        leastSigBits = leastSigBits or java.lang.Long.decode(s5).toLong()

        return UUID(mostSigBits, leastSigBits)
    }

    fun fromString(str: String): UUID {
        return if (str.indexOf('-') > 0) {
            fromStringWithHyphens(str)
        } else {
            fromStringWhitoutHyphens(str)
        }
    }

    fun toString(uuid: UUID): String {
        return uuid.toString()
    }

    fun toStringWhitoutHyphens(uuid: UUID): String {
        val msb = uuid.mostSignificantBits
        val lsb = uuid.leastSignificantBits
        return (digits(msb shr 32, 8) + digits(msb shr 16, 4) + digits(msb, 4)
                + digits(lsb shr 48, 4) + digits(lsb, 12))
    }

    private fun digits(`val`: Long, digits: Int): String {
        val hi = 1L shl digits * 4
        return java.lang.Long.toHexString(hi or (`val` and hi - 1)).substring(1)
    }

    fun fromByte(data: ByteArray): UUID {
        if (data.size != 16) {
            throw IllegalArgumentException("Invalid UUID byte[]")
        }
        var byteBuffer = ByteBuffer.wrap(data)
        var high = byteBuffer.long
        var low = byteBuffer.long
        return UUID(high, low)
    }

    fun toByte(uuid: UUID): ByteArray {
        val ba = ByteArrayOutputStream(16)
        val da = DataOutputStream(ba)
        try {
            da.writeLong(uuid.mostSignificantBits)
            da.writeLong(uuid.leastSignificantBits)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return ba.toByteArray()
    }
}
